product_dimension/models/product.py

```diff
@@ -6,76 +6,98 @@
 
-    product_length = fields.Float('length')
-    product_height = fields.Float('height')
-    product_width = fields.Float('width')
-    dimensional_uom_id = fields.Many2one(
-        'uom.uom',
-        'Dimensional UoM',
-        domain=lambda self: self._get_dimension_uom_domain(),
-        help='UoM for length, height, width',
-        default=lambda self: self.env.ref('uom.product_uom_meter'),
+    @api.onchange(
+        "product_length",
+        "product_height",
+        "product_width",
+        "dimensional_uom_id",
+        "volume_uom_id",
+    )
+    def onchange_calculate_volume(self):
+        self.volume = self.env["product.template"]._calc_volume(
+            self.product_length,
+            self.product_height,
+            self.product_width,
+            self.dimensional_uom_id,
+            self.volume_uom_id,
         )
 
 
 class ProductTemplate(models.Model):
-    _inherit = 'product.template'
+    _inherit = "product.template"
 
-    @api.model
-    def _calc_volume(self, product_length, product_height, product_width, uom_id):
-        volume = 0
-        if product_length and product_height and product_width and uom_id:
-            length_m = self.convert_to_meters(product_length, uom_id)
-            height_m = self.convert_to_meters(product_height, uom_id)
-            width_m = self.convert_to_meters(product_width, uom_id)
-            volume = length_m * height_m * width_m
+    def _convert_to_meters(self, measure, dimensional_uom):
+        uom_meters = self.env.ref("uom.product_uom_meter")
 
-        return volume
+        return dimensional_uom._compute_quantity(
+            qty=measure, to_unit=uom_meters, round=False,
+        )
 
-    @api.onchange(
-        'product_length', 'product_height', 'product_width', 'dimensional_uom_id')
-    def onchange_calculate_volume(self):
-        self.volume = self._calc_volume(
-            self.product_length, self.product_height, self.product_width,
-            self.dimensional_uom_id)
+    @api.model
+    def _calc_volume(
+        self, product_length, product_height, product_width, uom_id, volume_uom_id
+    ):
+        uom_litre = self.env.ref("uom.product_uom_litre")
+        volume_litre = 0
 
-    def convert_to_meters(self, measure, dimensional_uom):
-        uom_meters = self.env.ref('uom.product_uom_meter')
+        if product_length and product_height and product_width and uom_id:
+            length_m = self._convert_to_meters(product_length, uom_id)
+            height_m = self._convert_to_meters(product_height, uom_id)
+            width_m = self._convert_to_meters(product_width, uom_id)
+            volume_litre = (length_m * height_m * width_m) * 1000
 
-        return dimensional_uom._compute_quantity(
-            qty=measure,
-            to_unit=uom_meters,
-            round=False,
+        return uom_litre._compute_quantity(
+            qty=volume_litre, to_unit=volume_uom_id, round=False,
```

Comment:
```suggestion
        return uom_m³._compute_quantity(
            qty=volume_m³, to_unit=volume_uom_id, round=False,
```

Reply:
Hi @hparfr ! Thanks for the review !
In fact I used litres because it's the easiest UoM I found registered in native uom module datas :
https://github.com/odoo/odoo/blob/12.0/addons/uom/data/uom_data.xml#L90

In your suggestion I didn't catch where does your `uom_m³` comes from...

Reply:
It's not available in 12.0 as UoM by default in v12 but it's hardcoded in product https://github.com/odoo/odoo/blob/12.0/addons/product/models/product_template.py#L83
 
By the way it's available in 14.0 https://github.com/odoo/odoo/blob/14.0/addons/uom/data/uom_data.xml#L95

I have updated my suggestion. In that way, for the migration in v14 there will be less changes.

Reply:
Hmm... As I understand `volume_litre` and `volume_m³` as purely temporary variables to be able to make the conversion based on a known UoM, it looks like it doesn't really matter what's this temporary variable UoM...

But ok to use both `volume_litre` and `volume_m³` instead of only `volume_litre` if it helps for the 14.0 migration.